### ISSUE TITLE
Re-authentication fails due to missing project id in context

### ISSUE THREAD

**Issue by xrl** [gophercloud/gophercloud#255]
I am using a branch of kubernetes which moves away from the rackspace gophercloud and over to this more actively maintained gophercloud lib. The kubernetes health checks (a master scans nova for all running instances, making sure they're still present. same idea for cinder vols.) would fail after about 12 hours, owing to a re-auth-related failure.

I added some extra info to the reauth wrapping error:

```
 func (e ErrDefault400) Error() string {
-       return "Invalid request due to incorrect syntax or missing required parameters."
+       return fmt.Sprintf("Invalid request due to incorrect syntax or missing required parameters. URL: %v, Body: %v", e.ErrUnexpectedResponseCode.URL, string(e.ErrUnexpectedResponseCode.Body))
 }
```

and rebuilt kubernetes. Now I see these errors for cinder/nova:

Nova:
```
Successfully re-authenticated, but got error executing request: Invalid request due to incorrect syntax or missing required parameters. URL: https://my.cloud:8774/v2.1/cfef9feef46441a29ab31117690d352a/servers/detail?name=%5Ekube-node-03%24&status=ACTIVE, Body: {"badRequest": {"message": "Malformed request URL: URL's project_id 'cfef9feef46441a29ab31117690d352a' doesn't match Context's project_id 'None'", "code": 400}}
```

and

Cinder:
```
Attach volume "3f4d29d9-8407-4c2a-a435-645757d5d25c" to instance "8c97ee08-b345-462b-80eb-2a0033d9bd5e" failed with Successfully re-authenticated, but got error executing request: Invalid request due to incorrect syntax or missing required parameters. URL: https://my.cloud:8776/v2/cfef9feef46441a29ab31117690d352a/volumes/detail, Body: {"badRequest": {"message": "Malformed request url", "code": 400}}
```

The nova error is more useful and points out that a context (is that the same thing as a token?) doesn't specify the `project_id`.

Sounds like there is a bug in the re-auth code if it is generating bad tokens. The code generally works fine for the first 12 hours.

**Comment by jrperritt** [comment]
> Sounds like there is a bug in the re-auth code if it is generating bad tokens

Firstly, that would indeed be strange: the re-auth code is exactly the auth code. I've never had an issue re-authenticating, but I also don't use project ID.
Secondly, so far there's no proof that this library is the cause; this could just as well be your OpenStack environment or Kubernetes. Have you taken measures to rule out those possibilities?

Basically, I'm not sure this is a Gophercloud issue. Do you have the relevant OpenStack logs? If we can confirm this is a Gophercloud issue, I'll gladly try hunt it down and implement a fix for it.

**Comment by xrl** [comment]
@jrperritt I totally get that it could be issue on the kubernetes side or openstack. How can I gather request IDs from failed requests?

**Comment by jrperritt** [comment]
> How can I gather request IDs from failed requests?

I think `nova.conf` specifies the log file location for the Nova logs.

**Comment by xrl** [comment]
@jrperritt I was thinking more of how to get that request id out of the gophercloud client. That way I can go spelunking with more information on my end.

**Comment by jrperritt** [comment]
Oh, of course, sorry. I was thinking it might be feasible to just grep the logs for the error message.

For Gophercloud, I've just checked and noticed it's a bit more of an inconvenience at the moment:
The request ID is returned as a header (`X-Compute-Request-Id` for Nova). Each Gophercloud response has a header field, but for most operations it's not populated because, for most purposes, it doesn't contain useful information (except, of course, in debugging cases like this). If I have some time next week, I'll go through and add headers to all the responses. For now you can log the response headers by passing a custom `http.Client` to `ProviderClient.HTTPClient`. It is mentioned (with a linked example demonstrating re-auth limiting and logging) [here](https://github.com/gophercloud/gophercloud/blob/master/auth_options.go#L49).

Alternatively, if you know the specific operation, you can change a single function in your Gophercloud code:
As an example, take the [`servers.Get` operation](https://github.com/gophercloud/gophercloud/blob/master/openstack/compute/v2/servers/requests.go#L291). Changing it to
```go
func Get(client *gophercloud.ServiceClient, id string) (r GetResult) {
	resp, err := client.Get(getURL(client, id), &r.Body, &gophercloud.RequestOpts{
		OkCodes: []int{200, 203},
	})
        r.Err = err
        if resp != nil {
            r.Header = resp.Header    
        }
	return
}
```

should return the header in the response object:
```go
res, err := servers.Get(client, id)
reqid := res.Header.Get("X-Compute-Request-Id")
server, err := res.Extract()
```

**Comment by xrl** [comment]
I have been able to make a standalone proof of concept. Looks like `TenantId` was causing the issue: it was used in the first request to keystone but then the value was cleared by other code. Subsequent reauth code failed to provide a scope causing it to fail.

The following diff got me going:

```diff
diff --git a/vendor/github.com/gophercloud/gophercloud/auth_options.go b/vendor/github.com/gophercloud/gophercloud/auth_options.go
index 7a16131..65da13a 100644
--- a/vendor/github.com/gophercloud/gophercloud/auth_options.go
+++ b/vendor/github.com/gophercloud/gophercloud/auth_options.go
@@ -140,12 +140,13 @@ func (opts *AuthOptions) ToTokenV3CreateMap(scope map[string]interface{}) (map[s
        var req request

        // Test first for unrecognized arguments.
-       if opts.TenantID != "" {
-               return nil, ErrTenantIDProvided{}
-       }
-       if opts.TenantName != "" {
-               return nil, ErrTenantNameProvided{}
-       }

        if opts.Password == "" {
                if opts.TokenID != "" {
@@ -249,7 +250,6 @@ func (opts *AuthOptions) ToTokenV3CreateMap(scope map[string]interface{}) (map[s
 }

 func (opts *AuthOptions) ToTokenV3ScopeMap() (map[string]interface{}, error) {
-
        var scope struct {
                ProjectID   string
                ProjectName string
@@ -259,8 +259,8 @@ func (opts *AuthOptions) ToTokenV3ScopeMap() (map[string]interface{}, error) {

        if opts.TenantID != "" {
                scope.ProjectID = opts.TenantID
-               opts.TenantID = ""
-               opts.TenantName = ""
        } else {
                if opts.TenantName != "" {
                        scope.ProjectName = opts.TenantName
```

**Comment by bradbeam** [comment]
I'm seeing this issue in the above kubernetes issue ( https://github.com/kubernetes/kubernetes/issues/44461 ) and applying the patch in #314 addressed it.

**Comment by xrl** [comment]
@jrperritt can you take a look at this issue/PR? we're relying it for upstream use in Kubernetes and we need to get this in to kube releases.

**Comment by jtopjian** [comment]
@xrl Just a heads up that I'm going to be digging into this and hope to have an answer soon. Your patch in #314 is indeed simple, but I want to make sure it's the correct implementation. The standalone test you made will help a lot with this, so thank you for creating it.

Give me a day or three, but this is something I'm looking at :)

**Comment by jtopjian** [comment]
Yep, I see what's going on here.

> Looks like TenantId was causing the issue: it was used in the first request to keystone but then the value was cleared by other code. Subsequent reauth code failed to provide a scope causing it to fail.

This is exactly correct.

This issue is specific to Gophercloud. I was able to reproduce this in a "standard" Packstack environment which removes the possibility of an application or cloud provider bug. I have a very simple acceptance test that can trigger this error, too. It's based off of @xrl's `openstack-token-test`, but you don't even have to wait an hour. Any attempt to reauth will trigger the error.

My understanding is that `ToTokenV3CreateMap` and `ToTokenV3ScopeMap` are defined both in [`auth_options.go`](https://github.com/gophercloud/gophercloud/blob/master/auth_options.go) and in the [v3 token requests](https://github.com/gophercloud/gophercloud/blob/master/openstack/identity/v3/tokens/requests.go) for similar purposes. It's the slight difference that is triggering the error.

`auth_options.go` attempts to create a Scope by looking at the specified fields in `AuthOptions` and determine the best scope to go with. It then tries to tidy up by clearing some values after they have been placed in a Scope map. Since the values are erased, they are no longer available to be read the next time around. Contrast this with the v3 token request where one can explicitly define a Scope, so this problem never occurs.

Removing lines 262 and 263 [here](https://github.com/gophercloud/gophercloud/pull/314/files) fix this. But since those values are now never cleared, they're flagged for error during reauth, which is why 142-149 must also be removed.

So I agree with the proposed solution in #314.

I do appreciate the procedural checks that have gone into `auth_options.go` to make sure the auth values specified by the user conform to a valid identity request. I don't believe #314 is going to set precedence for more of that to go away. I agree that `TenantID` and `TenantName` should not be part of the `ToTokenCreateMap` since they're part of the scope, but they never make it to the actual POST _and_ they are essential to the purpose of the `auth_options.go`'s version of token creation.

Unless there are objections, I'll merge #314 tomorrow.

**Comment by jtopjian** [comment]
Fixed by #314 and #330
